Refactor twitch code

Daggersoath opened the pull request:
Moved all TwitchAPI related types into "TwitchAPI/apiTypes"
Replaced "request-promise" with "node-fetch"
Updated "updateChattersAwesomeness" argument to number
Created "Cache" class
Changed curStreamId to string
Changed messageQueueLimit to 1.5s to minimum limit
Added "tryCatchLog" to Logger to reduce duplicate code
yarn.lock updated
Info commands now have a rate limit of 10 minutes
Added trello command
Renamed api types to differentiate what is returned from the API
Rearranged endpoints to implement more one liners - sorry not sorry
Changed callTwitchApi options to headers object
More one liners in TwitchAPI
Less repeating "try -> do thing -> catch with log"
Fixed fetch calls to use proper options and specify headers

Commit:
All twitch related code moved from Util to TwitchAPI
Moved all TwitchAPI related types into "TwitchAPI/apiTypes"
Replaced "request-promise" with "node-fetch"
Updated "updateChattersAwesomeness" argument to number
Created "Cache" class
Changed curStreamId to string
Changed messageQueueLimit to 1.5s to minimum limit
Added "tryCatchLog" to Logger to reduce duplicate code
yarn.lock updated

Commit:
Info commands now have a rate limit of 10 minutes
Added trello command

Commit:
Renamed api types to differentiate what is returned from the API
Rearranged endpoints to implement more one liners
Changed callTwitchApi options to headers object
More one liners in TwitchAPI
Less repeating "try -> do thing -> catch with log"
Fixed fetch calls to use proper options and specify headers

scryptonite commented on src/beastie/commands/awesomeness/awesomeness.ts:
```diff
@@ -11,7 +11,7 @@ const execute = async (context: CommandContext): Promise<string> => {
   let twitchId: string = context.twitchId;
   if (context.para1) {
-    twitchId = await getTwitchId(context.para1);
+    twitchId = (await broadcaster.getProfile())?.id;
```
This line introduces a bug where you can't do `!awesomeness <username>` to retrieve the awesomeness for someone other than the broadcaster.

As your commit stands, `!awesomeness` will return whoever fired the command's awesomeness, while `!awesomeness SomeoneElse` will always return the broadcaster's awesomeness.

Perhaps you meant the following?
```suggestion
    twitchId = (await getTwitchProfile(context.para1))?.id;
```

*Should be batched with* https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429035796

scryptonite commented on src/beastie/index.ts:
```diff
@@ -141,7 +142,7 @@ export default class BeastieBot {
 
     this.state.isStreaming = response.live;
-    this.state.curStreamId = response.streamId;
+    this.state.curStreamId = response.streamId.toString();
```
Are you calling `toString()` because the `streamId` is of type `any`? I presume it's a real string when it is in the response, but that this call is just for satisfying the type-checker.

Perhaps the following will satisfy the type-checker without redundantly calling `.toString()`?
```suggestion
    this.state.curStreamId = response.streamId as string;
```

Daggersoath replied:
streamId is a number currently, all this comes from twitchWebhooks so needs to be refactored and fixed at a later point.
Going to leave it as is now and fix it when we fix webhooks

scryptonite commented on src/services/twitch/TwitchAPI/apiTypes.ts:
```diff
@@ -0,0 +1,51 @@
+
+export type TwitchStreamsReturnData = {
+  data: TwitchStream[];
+  pagination: any; // idk what this is but twitch includes it in the response (never seen it with any data)
```
ℹ️ With regards to this comment, the pagination property may contain a `cursor` that can be used to walk the streams that match the query supplied to the endpoint, if the count of streams exceed the supplied or default maximum number of stream objects you've queried for. (Useful for being able to iterate every stream in a particular game category, for example).

➡️ https://dev.twitch.tv/docs/api/reference#get-streams

*No suggestion given, just sharing an FYI :)*

scryptonite commented on src/utils/index.ts:
```diff
@@ -1,196 +1,64 @@
+} from "../services/twitch/TwitchAPI";
+import { default as fetch, RequestInfo } from "node-fetch";
+
+export async function PerformGetRequest<T>(
```
I think that because `PerformGetRequest` is not a class or constructor, the first character should be lowercase.

```suggestion
export async function performGetRequest<T>(
```
- *Should be batched with*
  - https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429037016
  - https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429036984

scryptonite commented on src/utils/Logging.ts:
```diff
@@ -44,4 +44,17 @@ if (process.env.NODE_ENV !== "production") {
   );
 }
 
+export async function tryCatchLog<T>(
+  func: () => Promise<T>,
+  logFunction,
+  outputOnFail
+): Promise<T> {
+  try {
+    return await func();
+  } catch (e) {
+    logFunction(`${outputOnFail}: ${e}`);
+  }
+  return null;
+}
+
```
Proposing that this could be re-written to work like the following:
```suggestion
export function swallowRejection(
    errorMessage: string,
    errorLog: (message: string) => void = BeastieLogger.error
): (rejection: any) => Promise<null> {
    return async (error: any) => {
        await errorLog(`${errorMessage}: ${JSON.stringify(error)}`);
        return null;
    };
}
```


Then for every promise that you want to "swallow" its possible rejection or error, you just tack on `.catch(swallowRejection("Thing broke"))` to the end of it, like so:

```js
// - mock example:
async function chanceOfRejection(threshold) {
    if (Math.random() > threshold)
        throw 123;
    return 456;
}

// ...
// - so instead of:
const nilOrNumber = await tryCatchLog(() => chanceOfRejection(0.5), "chanceOfRejection error", console.error));
// - you would:
const nilOrNumber = await chanceOfRejection(0.5).catch(swallowRejection("chanceOfRejection error", console.error));
```

If you accept this suggestion, don't forget to update the references. :) If you have questions or thoughts about what I've proposed, feel free to let me know.

scryptonite commented on src/services/twitch/TwitchAPI/index.ts:
```diff
@@ -0,0 +1,89 @@
+// query.
+import { Cache, CacheKeyValPair } from "../../../utils/cache";
+import { TwitchProfile, TwitchStream } from "./apiTypes";
+import { BeastieLogger, tryCatchLog } from "../../../utils/Logging";
```
```suggestion
import { BeastieLogger, swallowRejection } from "../../../utils/Logging";
```
See ➡️ https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429028595

scryptonite commented on src/services/twitch/TwitchAPI/index.ts:
```diff
@@ -0,0 +1,89 @@
+export const getTwitchProfile = async (
+  username: string
+): Promise<TwitchProfile> =>
+  tryCatchLog(
+    async () => (await profileCache.get(username)) as TwitchProfile,
+    BeastieLogger.warn,
+    `Failed to get twitch user ${username}`
+  );
```
```suggestion
  (profileCache.get(username) as Promise<TwitchProfile>)
    .catch(swallowRejection(`Failed to get twitch user ${username}`, BeastieLogger.warn));
```
See ➡️ https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429028595

scryptonite commented on src/services/twitch/TwitchAPI/index.ts:
```diff
@@ -0,0 +1,89 @@
+export const getTwitchProfiles = async (
+  usernames: string[]
+): Promise<TwitchProfile[]> =>
+  tryCatchLog<TwitchProfile[]>(
+    async () => (await profileCache.get(usernames)) as TwitchProfile[],
+    BeastieLogger.warn,
+    `Failed to get twitch user ${usernames}`
+  );
```
```suggestion
  (profileCache.get(usernames) as Promise<TwitchProfile[]>)
    .catch(swallowRejection(`Failed to get twitch user ${usernames}`, BeastieLogger.warn));
```

See ➡️ https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429028595

scryptonite commented on src/utils/index.ts:
```diff
@@ -1,196 +1,64 @@
+    return `I don't think that is a number o.o`;
   }
+
+  let viewers: TwitchProfile[] = await tryCatchLog<TwitchProfile[]>(
+    async () => getTwitchChatters(config.BROADCASTER_USERNAME),
+    BeastieLogger.warn,
+    "Failed to get chatroom viewers"
+  );
```
```suggestion
  let viewers: TwitchProfile[] = await (getTwitchChatters(config.BROADCASTER_USERNAME) as Promise<TwitchProfile[]>)
    .catch(swallowRejection("Failed to get chatroom viewers", BeastieLogger.warn));
```
See ➡️ https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429028595

scryptonite commented on src/utils/index.ts:
```diff
@@ -1,196 +1,64 @@
+  user,
+  amount
+): Promise<string> => {
+  let profile: TwitchProfile = await tryCatchLog<TwitchProfile>(
+    async () => getTwitchProfile(user),
+    BeastieLogger.warn,
+    `Failed to get twitch profile for ${user}`
+  );
```
```suggestion
  let profile: TwitchProfile = await (getTwitchProfile(user) as Promise<TwitchProfile>)
    .catch(swallowRejection(`Failed to get twitch profile for ${user}`, BeastieLogger.warn));
```
See ➡️ https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429028595

scryptonite commented on src/utils/index.ts:
```diff
@@ -1,196 +1,64 @@
-import rp from "request-promise";
 import config from "../config";
-import { BeastieLogger } from "./Logging";
+import { BeastieLogger, tryCatchLog } from "./Logging";
```
```suggestion
import { BeastieLogger, swallowRejection } from "./Logging";
```
See ➡️ https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429028595

scryptonite committed:
Suggested bulk edit to tryCatchBlock semantics

Instead of making a function to call and wrap a promise to catch and log a possible error, we create something of a higher order function that can be used as a rejection handler via any given promise's `.catch` method.

scryptonite commented on src/beastie/commands/awesomeness/awesomeness.ts:
```diff
@@ -1,8 +1,8 @@
 import { CommandModule } from "../index";
 import CommandContext from "../utils/commandContext";
 import { getAwesomeness } from "../../../services/db";
+import { broadcaster } from "../../../services/twitch/TwitchAPI";
```
*Associated with* ➡️ https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429017224
```suggestion
import { getTwitchProfile } from "../../../services/twitch/TwitchAPI";
```

scryptonite commented on src/services/twitch/TwitchAPI/endpoints.ts:
```diff
@@ -0,0 +1,43 @@
+    },
+    ...headers
+  };
+  return PerformGetRequest<T>(uri, httpHeaders);
```
```suggestion
  return performGetRequest<T>(uri, httpHeaders);
```
*Batched with* https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429022490

scryptonite commented on src/services/twitch/TwitchAPI/endpoints.ts:
```diff
@@ -0,0 +1,43 @@
+  TwitchUsersReturnData
+} from "./apiTypes";
+import config from "../../../config";
+import { PerformGetRequest } from "../../../utils";
```
```suggestion
import { performGetRequest } from "../../../utils";
```
*Batched with* https://github.com/teamTALIMA/BeastieBot-V3/pull/13/files#r429022490

scryptonite commented:
Looks like another good PR :) 

Sorry about committing the `tryCatchLog`➡️`swallowRejection` changes... I was using a part of GitHub I hadn't before to try to make it easier for you to accept a batch of suggestions.

Let me know what you think of my review suggestions.

Daggersoath committed:
Fix bug where awesomeness would only check broadcaster rather than the parameter sent in

Co-authored-by: Scryptonite <scryptonite.mail@gmail.com>

Daggersoath committed:
Renaming PerformGetRequest to camelCase to fit with coding standard

Co-authored-by: Scryptonite <scryptonite.mail@gmail.com>

Daggersoath committed:
performGetRequest

Co-authored-by: Scryptonite <scryptonite.mail@gmail.com>

Daggersoath committed:
performGetRequest

Co-authored-by: Scryptonite <scryptonite.mail@gmail.com>

Daggersoath committed:
replacing broadcaster with getTwitchProfile to fix bug where only broadcasters awesomeness could be fetched

Co-authored-by: Scryptonite <scryptonite.mail@gmail.com>

Commit:
Fixed typo in Logging BeastieLogger -> beastieLogger
Updated comment in apiTypes

Daggersoath reviewed (commented):
Went through Scryptonite's changes, love them :)

scryptonite reviewed (approved):
I approve of these changes. Thanks again 👍
